ntoskrnl/io/iomgr/deviface.c

```diff
@@ -30,185 +31,437 @@ static PWCHAR BaseKeyString = L"\\Registry\\Machine\\System\\CurrentControlSet\\
 
 static
 NTSTATUS
-OpenRegistryHandlesFromSymbolicLink(IN PUNICODE_STRING SymbolicLinkName,
-                                    IN ACCESS_MASK DesiredAccess,
-                                    IN OPTIONAL PHANDLE GuidKey,
-                                    IN OPTIONAL PHANDLE DeviceKey,
-                                    IN OPTIONAL PHANDLE InstanceKey)
+IopBuildSymbolicLink(_In_ PUNICODE_STRING PrefixString,
+                     _In_ PUNICODE_STRING DeviceString,
+                     _In_ PUNICODE_STRING GuidString,
+                     _In_opt_ PUNICODE_STRING ReferenceString,
+                     _Out_ PUNICODE_STRING SymbolicLinkName)
 {
-    OBJECT_ATTRIBUTES ObjectAttributes;
-    UNICODE_STRING BaseKeyU;
-    UNICODE_STRING GuidString, SubKeyName, ReferenceString;
-    PWCHAR StartPosition, EndPosition;
-    HANDLE ClassesKey;
-    PHANDLE GuidKeyRealP, DeviceKeyRealP, InstanceKeyRealP;
-    HANDLE GuidKeyReal, DeviceKeyReal, InstanceKeyReal;
+    UNICODE_STRING SymbolicLink;
     NTSTATUS Status;
 
-    SubKeyName.Buffer = NULL;
+    /* Build up new symbolic link */
+    SymbolicLink.Length = 0;
+    SymbolicLink.MaximumLength = PrefixString->Length + DeviceString->Length + sizeof(L"#") + GuidString->Length + (ReferenceString != NULL ? ReferenceString->Length : sizeof(L"\\")) + sizeof(UNICODE_NULL);
+    SymbolicLink.Buffer = ExAllocatePoolWithTag(PagedPool, SymbolicLink.MaximumLength, TAG_IO);
+    if (!SymbolicLink.Buffer)
+    {
+        DPRINT1("ExAllocatePoolWithTag() failed\n");
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    Status = RtlAppendUnicodeStringToString(&SymbolicLink, PrefixString);
```

Comment:
Looking at all this sequence of functions, I'm wondering whether it could be interesting to sum them up into one single printf-like call instead:
```c
    Status = RtlUnicodeStringPrintf(&SymbolicLink,
                                    L"%wZ%wZ#%wZ%wZ",
                                    PrefixString,
                                    DeviceString,
                                    GuidString,
                                    ReferenceString);
    NT_ASSERT(NT_SUCCESS(Status));
```
(Note also that I have used my previous comment regarding `ReferenceString` in order to be able to use it unconditionally here.)

Reply:
Thank you, applied your suggestion. I did not even know about this function! It much more simplifies the code! :smile: 
Also applied a reference string suggestion.